File: long_questions/DSA/Leetcode/golang/2_Hashing_Frequency_Map/217_Contains_Duplicate.go
package main

import "fmt"

// 217. Contains Duplicate
// Time: O(N), Space: O(N)
func containsDuplicate(nums []int) bool {
	numSet := make(map[int]bool)
	
	for _, num := range nums {
		if _, exists := numSet[num]; exists {
			return true
		}
		numSet[num] = true
	}
	
	return false
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Hash Set for Duplicate Detection
- **Hash Set Usage**: Track seen elements efficiently
- **Early Termination**: Return immediately when duplicate found
- **Constant Time Lookup**: O(1) average time for membership test
- **Space-Time Tradeoff**: Use extra space for O(N) time

## 2. PROBLEM CHARACTERISTICS
- **Duplicate Detection**: Check if any element appears more than once
- **Existence Query**: Only need to know if duplicate exists, not where
- **Early Exit**: Can stop as soon as first duplicate is found
- **Set Natural Fit**: Hash set perfect for tracking seen elements

## 3. SIMILAR PROBLEMS
- Contains Duplicate II (LeetCode 219) - Duplicates within distance k
- Contains Duplicate III (LeetCode 220) - Duplicates within distance and value range
- First Missing Positive (LeetCode 41) - Find smallest missing positive
- Find Duplicate Number (LeetCode 287) - Find duplicate in array

## 4. KEY OBSERVATIONS
- **Hash set ideal**: O(1) average lookup and insertion
- **Early termination**: Can stop when first duplicate found
- **Space requirement**: Need to store seen elements
- **Linear scan**: Single pass through array sufficient

## 5. VARIATIONS & EXTENSIONS
- **Distance constraint**: Duplicates within specific distance
- **Value range constraint**: Duplicates within specific value difference
- **Multiple duplicates**: Count all duplicates or find all duplicate pairs
- **Streaming data**: Handle data that arrives over time

## 6. INTERVIEW INSIGHTS
- Always clarify: "Can I modify the input? What about space constraints?"
- Edge cases: empty array (false), single element (false), all same (true)
- Time complexity: O(N) time, O(N) space
- Alternative: O(N log N) time, O(1) space with sorting

## 7. COMMON MISTAKES
- Using nested loops (O(N²) time)
- Not handling empty array case
- Using array instead of hash set (inefficient lookups)
- Not returning early when duplicate found
- Forgetting negative numbers or zero

## 8. OPTIMIZATION STRATEGIES
- **Early termination**: Stop when first duplicate found
- **Pre-size hash set**: Estimate capacity to reduce reallocations
- **Bit set optimization**: For limited integer ranges
- **Sorting alternative**: If space is constrained

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like checking for duplicate tickets:**
- You're checking tickets (numbers) at an event entrance
- You want to know if anyone has a duplicate ticket
- You keep a list of tickets you've already seen
- For each new ticket, you check if it's already in your list
- If you find a duplicate, you can stop checking immediately

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of integers
2. **Goal**: Determine if any element appears more than once
3. **Output**: Boolean (true if duplicate exists, false otherwise)
4. **Constraint**: Need efficient detection

#### Phase 2: Key Insight Recognition
- **"Hash set natural fit"** → Perfect for tracking seen elements
- **"Early termination"** → Can stop when first duplicate found
- **"Linear scan sufficient"** → No need for nested loops
- **"Space-time tradeoff"** → Use extra space for O(N) time

#### Phase 3: Strategy Development
```
Human thought process:
"I need to check if any number appears more than once.
I'll keep track of numbers I've seen using a hash set.
For each number, I'll check if it's already in the set.
If it is, I found a duplicate and can return true immediately.
If not, I'll add it to the set and continue.
If I finish the array without finding duplicates, return false."
```

#### Phase 4: Edge Case Handling
- **Empty array**: No duplicates possible, return false
- **Single element**: No duplicates possible, return false
- **All same elements**: First duplicate found at second element
- **Large arrays**: Handle efficiently with hash set

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Input: [1, 2, 3, 1]

Human thinking:
"I'll process each number and track what I've seen:

Number 1:
- Check if 1 is in seen set: No
- Add 1 to seen set: {1}

Number 2:
- Check if 2 is in seen set: No
- Add 2 to seen set: {1, 2}

Number 3:
- Check if 3 is in seen set: No
- Add 3 to seen set: {1, 2, 3}

Number 1:
- Check if 1 is in seen set: Yes!
- Found duplicate, return true immediately

No need to process further!"
```

#### Phase 6: Intuition Validation
- **Why hash set works**: O(1) average lookup and insertion
- **Why early termination**: Finding any duplicate is sufficient
- **Why O(N) time**: Single pass through array
- **Why O(N) space**: Need to store up to N elements

### Common Human Pitfalls & How to Avoid Them
1. **"Why not use sorting?"** → Sorting is O(N log N) and modifies input
2. **"Should I use nested loops?"** → O(N²) time, too slow for large inputs
3. **"What about space constraints?"** → Sorting alternative if space limited
4. **"Can I optimize further?"** → Early termination is already optimal

### Real-World Analogy
**Like checking for duplicate IDs at a conference:**
- You're registering attendees and checking their ID numbers
- You want to ensure no two attendees have the same ID
- You keep a list of IDs you've already registered
- For each new attendee, you check if their ID is already in your list
- If you find a duplicate, you flag it immediately

### Human-Readable Pseudocode
```
function containsDuplicate(nums):
    seenSet = set()
    
    for num in nums:
        if num in seenSet:
            return true
        seenSet.add(num)
    
    return false
```

### Execution Visualization

### Example: [1, 2, 3, 1]
```
Hash Set Evolution:
Processing 1:
- 1 not in {} → add 1
- set = {1}

Processing 2:
- 2 not in {1} → add 2
- set = {1, 2}

Processing 3:
- 3 not in {1, 2} → add 3
- set = {1, 2, 3}

Processing 1:
- 1 is in {1, 2, 3} → FOUND DUPLICATE!
- Return true immediately

Result: true
```

### Key Visualization Points:
- **Hash set tracking**: Keep track of seen elements
- **Constant time lookup**: O(1) average time for membership test
- **Early termination**: Stop as soon as duplicate found
- **Linear progression**: Single pass through array

### Memory Layout Visualization:
```
Array Processing Flow:
[1, 2, 3, 1]
  ↓
{1}
  ↓
{1, 2}
  ↓
{1, 2, 3}
  ↓
{1, 2, 3} ← 1 already exists → DUPLICATE FOUND!

Hash Set Contents:
Step 1: {1}
Step 2: {1, 2}
Step 3: {1, 2, 3}
Step 4: {1, 2, 3} (duplicate detected)
```

### Time Complexity Breakdown:
- **Hash set approach**: O(N) time, O(N) space
- **Sorting approach**: O(N log N) time, O(1) space (if in-place)
- **Nested loops**: O(N²) time, O(1) space (too slow)
- **Bit set**: O(N) time, O(R) space where R is value range

### Alternative Approaches:

#### 1. Sorting Approach (O(N log N) time, O(1) space)
```go
func containsDuplicateSort(nums []int) bool {
    sort.Ints(nums)
    
    for i := 1; i < len(nums); i++ {
        if nums[i] == nums[i-1] {
            return true
        }
    }
    
    return false
}
```
- **Pros**: O(1) extra space, no hash map overhead
- **Cons**: O(N log N) time, modifies input array

#### 2. Bit Set for Limited Range (O(N) time, O(R) space)
```go
func containsDuplicateBitSet(nums []int, maxValue int) bool {
    if maxValue > 1000000 { // Only for limited ranges
        return containsDuplicate(nums) // Fall back to hash set
    }
    
    bitSet := make([]uint64, (maxValue/64)+1)
    
    for _, num := range nums {
        if num < 0 {
            continue // Handle negative numbers separately
        }
        
        index := num / 64
        bit := uint64(1) << (num % 64)
        
        if bitSet[index]&bit != 0 {
            return true
        }
        bitSet[index] |= bit
    }
    
    return false
}
```
- **Pros**: Very memory efficient for limited ranges
- **Cons**: Only works for non-negative integers in limited range

#### 3. Floyd's Tortoise and Hare (for specific constraints)
```go
// Only works if numbers are in range [1, n-1] and array length is n
func containsDuplicateFloyd(nums []int) bool {
    // This is for Find Duplicate Number problem, not general case
    // Included here for educational purposes
    return false // Not applicable for general duplicate detection
}
```
- **Pros**: O(1) space for specific constraints
- **Cons**: Only works for very specific problem variants

### Extensions for Interviews:
- **Contains Duplicate II**: Duplicates within distance k
- **Contains Duplicate III**: Duplicates within distance and value range
- **Find All Duplicates**: Return all elements that appear more than once
- **Streaming Data**: Handle data that arrives over time
- **Count Duplicates**: Count how many duplicates exist
*/
func main() {
	// Test cases
	testCases := [][]int{
		{1, 2, 3, 1},
		{1, 2, 3, 4},
		{1, 1, 1, 3, 2, 2, 2},
		{},
		{0},
		{-1, -2, -3, -1},
	}
	
	for i, nums := range testCases {
		result := containsDuplicate(nums)
		fmt.Printf("Test Case %d: %v -> Contains duplicate: %t\n", i+1, nums, result)
	}
}

File: long_questions/DSA/Leetcode/java/2_Hashing_Frequency_Map/217_Contains_Duplicate.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class ContainsDuplicate {
    
    // 217. Contains Duplicate
    // Time: O(N), Space: O(N)
    public static boolean containsDuplicate(int[] nums) {
        Set<Integer> numSet = new HashSet<>();
        
        for (int num : nums) {
            if (!numSet.add(num)) {
                return true;
            }
        }
        
        return false;
    }

    public static void main(String[] args) {
        // Test cases
        int[][] testCases = {
            {1, 2, 3, 1},
            {1, 2, 3, 4},
            {1, 1, 1, 3, 2, 2, 2},
            {},
            {0},
            {-1, -2, -3, -1}
        };
        
        for (int i = 0; i < testCases.length; i++) {
            boolean result = containsDuplicate(testCases[i]);
            System.out.printf("Test Case %d: %s -> Contains duplicate: %b\n", 
                i + 1, Arrays.toString(testCases[i]), result);
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Hash Set for Duplicate Detection
- **Hash Set**: Stores unique elements encountered
- **Early Detection**: Return true as soon as duplicate found
- **Single Pass**: Process array once
- **Set Properties**: add() returns false for duplicates

## 2. PROBLEM CHARACTERISTICS
- **Array Input**: Collection of integers
- **Duplicate Check**: Determine if any element appears twice
- **Boolean Result**: Simple true/false answer
- **Early Exit**: Can stop as soon as duplicate found

## 3. SIMILAR PROBLEMS
- Contains Duplicate II (within k distance)
- Contains Duplicate III (within t value)
- Find All Duplicates in Array
- Single Number

## 4. KEY OBSERVATIONS
- Hash set provides O(1) duplicate detection
- Set.add() returns false if element already exists
- Can return immediately when duplicate found
- No need to process entire array if duplicate found early

## 5. VARIATIONS & EXTENSIONS
- Find all duplicates
- Count occurrences of each element
- Handle distance constraints (k apart)
- Handle value difference constraints (t apart)

## 6. INTERVIEW INSIGHTS
- Clarify: "What counts as duplicate?" (exact same value)
- Edge cases: empty array, single element, all unique
- Space-time tradeoff: O(N) space vs O(N²) time
- Alternative: Sort first, then check adjacent

## 7. COMMON MISTAKES
- Using array instead of set (inefficient)
- Not returning early when duplicate found
- Mishandling empty array case
- Forgetting about hash set properties

## 8. OPTIMIZATION STRATEGIES
- Current solution is optimal for time
- For memory constraints, consider sorting approach
- For streaming data, use bounded-size set
- For range-limited data, use boolean array

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like tracking party guests:**
- You're checking IDs at a party entrance
- You keep a list of everyone who has entered
- When someone new arrives, you check if they're already on your list
- If yes, they're a duplicate - party crasher!
- If no, you add them to your list

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Array of integers
2. **Goal**: Check if any number appears more than once
3. **Output**: True if duplicate exists, false otherwise

#### Phase 2: Key Insight Recognition
- **"I need to remember what I've seen!"** → Use a set
- **"How to check quickly?"** → Hash set O(1) lookup
- **"When to stop?"** → As soon as duplicate found
- **"Set.add() is perfect!"** → Returns false for duplicates

#### Phase 3: Strategy Development
```
Human thought process:
"I'll create an empty set to track unique numbers.
For each number, I'll try to add it to the set.
If add() returns false, it means I've seen it before - duplicate!
If I make it through the whole array, no duplicates found."
```

#### Phase 4: Edge Case Handling
- **Empty array**: No duplicates, return false
- **Single element**: Cannot have duplicate, return false
- **All same elements**: First duplicate at index 1
- **All unique**: Process entire array, return false

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Starting with: [1, 2, 3, 1]

Human thinking:
"Let me create my empty set.
Position 0: number = 1, set = {} → add 1 successfully
Position 1: number = 2, set = {1} → add 2 successfully  
Position 2: number = 3, set = {1,2} → add 3 successfully
Position 3: number = 1, set = {1,2,3} → try to add 1
           1 is already in set! Duplicate found!
           I can stop right here and return true."
```

#### Phase 6: Intuition Validation
- **Why it works**: Set automatically handles uniqueness
- **Why it's efficient**: O(1) operations, early exit possible
- **Why it's correct**: We check every element against all previous ones

### Common Human Pitfalls & How to Avoid Them
1. **"Should I use an array?"** → No, set is much faster
2. **"Can I optimize space?"** → Only with sorting (slower)
3. **"What about early exit?"** → Always return as soon as duplicate found
4. **"How to handle negatives?"** → Set handles them fine

### Real-World Analogy
**Like checking for duplicate library books:**
- You're scanning books into a library system
- You keep a record of all book IDs you've scanned
- When you scan a new book, you check if it's already in your system
- If yes, someone's trying to return the same book twice!
- If no, you add it to your system and continue

### Human-Readable Pseudocode
```
function containsDuplicate(numbers):
    seenNumbers = empty set
    
    for each number in numbers:
        if number cannot be added to seenNumbers:
            return true  // duplicate found
        add number to seenNumbers
    
    return false  // no duplicates found
```

### Execution Visualization

### Example: [1, 2, 3, 1]
```
Initial: nums = [1, 2, 3, 1], set = {}

Step 1: i=0, nums[0]=1
→ add(1) succeeds
→ set = {1}
State: set={1}, result=false

Step 2: i=1, nums[1]=2
→ add(2) succeeds
→ set = {1, 2}
State: set={1,2}, result=false

Step 3: i=2, nums[2]=3
→ add(3) succeeds
→ set = {1, 2, 3}
State: set={1,2,3}, result=false

Step 4: i=3, nums[3]=1
→ add(1) fails (already exists)
→ Return true immediately
State: set={1,2,3}, result=true
```

### Key Visualization Points:
- **Hash set** stores unique elements only
- **add() operation** fails for duplicates
- **Early return** when duplicate detected
- **Single pass** through array

### Memory Layout Visualization:
```
Array: [1][2][3][1]
       ^
       |
       +-- Current: 1 (duplicate!)
       
Set Evolution:
Step 0: {}
Step 1: {1}
Step 2: {1,2}
Step 3: {1,2,3}
Step 4: {1,2,3} ✓ Duplicate found!
```

### Time Complexity Breakdown:
- **Best Case**: O(1) - duplicate found at index 1
- **Worst Case**: O(N) - no duplicates, process all elements
- **Average Case**: O(N) - typical scenario
- **Space**: O(N) - storing unique elements
*/
